### PR TITLE
retrieve all entries in batches to prevent OOM

### PR DESCRIPTION
**Pull request by JimWen**
raft: retrieve all log entries in batches to prevent OOM

Trying to retrieve all entries at once might cause out-of-memory issues in
cases where the raft log is too big to fit into memory. Instead of retrieving
all entries at once, retrieve it in batches of 64MB.

It's a problem caught in dgraph which use raft and oom occurs when WAL log is too large.
https://discuss.dgraph.io/t/dgraph-cant-idle-without-being-oomkilled-after-large-data-ingestion/6543/56

Code modify from a commit to solve a similar problem caught in dgraph before and the solution has worked fine for a long time.
https://github.com/dgraph-io/dgraph/commit/69d9403898548327360fc285dbc6965177d33ea4

### CODE REVIEW

**Comment by danielmai** [comment]
@JimWen also shared a heap profile analysis here: https://discuss.dgraph.io/t/dgraph-cant-idle-without-being-oomkilled-after-large-data-ingestion/6543/60

File: raft/raft.go [hunk 1]
```diff
@@ -911,11 +911,31 @@ func (r *raft) Step(m pb.Message) error {
...
-			if err != nil {
-				r.logger.Panicf("unexpected error getting unapplied entries (%v)", err)
+
+			n := 0
+
+			// Trying to retrieve all entries at once might cause out-of-memory issues in
+			// cases where the raft log is too big to fit into memory. Instead of retrieving
+			// all entries at once, retrieve it in batches of 64MB.
+			for batchFirst := r.raftLog.applied + 1; batchFirst <= r.raftLog.committed; {
+				ents, err := r.raftLog.slice(batchFirst, r.raftLog.committed+1, 64<<20)
```
Comment (gyuho) [inline raft/raft.go:921]: Can we define `64<<20` as a constant with comment?

**Comment by gyuho** [comment]
Can any of @bdarnell @tbg @xiang90 comment on this?

File: raft/raft.go (hunk 1 above)
Comment (bdarnell) [reply raft/raft.go:921]: Maybe this should use Config.MaxCommittedSizePerReady instead.

Comment (JimWen) [reply raft/raft.go:921]: > Maybe this should use Config.MaxCommittedSizePerReady instead.

Config.MaxCommittedSizePerReady is set to 1 * 1024 * 1024(1m) default, that may be a little small batch.

**Pull request closed without merging**